### ISSUE TITLE
read_chunk does not accept non-literal input

### ISSUE THREAD

**Issue by lbartnik** [yihui/knitr#1647]
Providing run-time calculated paths used to work in RMarkdown vignettes.

```
knitr::read_chunk(system.file('scripts/iris-model.R', package = 'repository'))
```

but after upgrading knitr and its dependencies to the latest form CRAN, that snippet throws an exception:

```
Error in parse(text = code, keep.source = FALSE) :
  <text>:2:0: unexpected end of input
1: read_chunk(system.file('scripts/iris-model.R', package = 'repository')
   ^
```

If I provide the literal path it compiles correctly:

```
knitr::read_chunk("/home/lukasz/R/x86_64-pc-linux-gnu-library/3.5/repository/scripts/iris-model.R")
```

I see that in `knitr/R/block.R`, line 473, there is a recent change that probably causes this error.

Session info:

```
devtools::session_info()
─ Session info ──────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 3.5.1 (2018-07-02)
 os       Ubuntu 18.10                
 system   x86_64, linux-gnu           
 ui       RStudio                     
 language (EN)                        
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       America/Los_Angeles         
 date     2018-12-13                  

─ Packages ──────────────────────────────────────────────────────────────────
 package     * version    date       lib source                           
 assertthat    0.2.0      2017-04-11 [1] CRAN (R 3.5.1)                   
 backports     1.1.2      2017-12-13 [1] CRAN (R 3.5.1)                   
 base64enc     0.1-3      2015-07-28 [1] CRAN (R 3.5.1)                   
 bindr         0.1.1      2018-03-13 [1] CRAN (R 3.5.1)                   
 bindrcpp      0.2.2      2018-03-29 [1] CRAN (R 3.5.1)                   
 broom         0.5.0      2018-07-17 [1] CRAN (R 3.5.1)                   
 browser       0.1        2018-10-07 [1] Github (lbartnik/browser@adc8b8d)
 callr         3.0.0      2018-08-24 [1] CRAN (R 3.5.1)                   
 chronicler    0.1.0      2018-12-13 [1] local                            
 cli           1.0.1      2018-09-25 [1] CRAN (R 3.5.1)                   
 crayon        1.3.4      2017-09-16 [1] CRAN (R 3.5.1)                   
 desc          1.2.0      2018-05-01 [1] CRAN (R 3.5.1)                   
 devtools      2.0.1      2018-10-26 [1] CRAN (R 3.5.1)                   
 digest        0.6.18     2018-10-10 [1] CRAN (R 3.5.1)                   
 dplyr         0.7.7      2018-10-16 [1] CRAN (R 3.5.1)                   
 evaluate      0.12       2018-10-09 [1] CRAN (R 3.5.1)                   
 fs            1.2.6      2018-08-23 [1] CRAN (R 3.5.1)                   
 glue          1.3.0      2018-07-17 [1] CRAN (R 3.5.1)                   
 htmltools     0.3.6      2017-04-28 [1] CRAN (R 3.5.1)                   
 httpuv        1.4.5      2018-07-19 [1] CRAN (R 3.5.1)                   
 jsonlite      1.5        2017-06-01 [1] CRAN (R 3.5.1)                   
 knitr         1.21       2018-12-10 [1] CRAN (R 3.5.1)                   
 later         0.7.5      2018-09-18 [1] CRAN (R 3.5.1)                   
 lattice       0.20-38    2018-11-04 [4] CRAN (R 3.5.1)                   
 lubridate     1.7.4      2018-04-11 [1] CRAN (R 3.5.1)                   
 magrittr      1.5        2014-11-22 [1] CRAN (R 3.5.1)                   
 memoise       1.1.0      2017-04-21 [1] CRAN (R 3.5.1)                   
 mime          0.6        2018-10-05 [1] CRAN (R 3.5.1)                   
 miniUI        0.1.1.1    2018-05-18 [1] CRAN (R 3.5.1)                   
 nlme          3.1-137    2018-04-07 [4] CRAN (R 3.5.0)                   
 packrat       0.5.0      2018-11-14 [1] CRAN (R 3.5.1)                   
 pillar        1.3.0      2018-07-14 [1] CRAN (R 3.5.1)                   
 pkgbuild      1.0.2      2018-10-16 [1] CRAN (R 3.5.1)                   
 pkgconfig     2.0.2      2018-08-16 [1] CRAN (R 3.5.1)                   
 pkgload       1.0.2      2018-10-29 [1] CRAN (R 3.5.1)                   
 png           0.1-7      2013-12-03 [1] CRAN (R 3.5.1)                   
 prettyunits   1.0.2      2015-07-13 [1] CRAN (R 3.5.1)                   
 processx      3.2.0      2018-08-16 [1] CRAN (R 3.5.1)                   
 promises      1.0.1      2018-04-13 [1] CRAN (R 3.5.1)                   
 proto         1.0.0      2016-10-29 [1] CRAN (R 3.5.1)                   
 ps            1.2.0      2018-10-16 [1] CRAN (R 3.5.1)                   
 purrr         0.2.5      2018-05-29 [1] CRAN (R 3.5.1)                   
 R6            2.3.0      2018-10-04 [1] CRAN (R 3.5.1)                   
 Rcpp          1.0.0      2018-11-07 [1] CRAN (R 3.5.1)                   
 remotes       2.0.2      2018-10-30 [1] CRAN (R 3.5.1)                   
 repository    0.1        2018-12-13 [1] local                            
 rlang         0.3.0.1    2018-10-25 [1] CRAN (R 3.5.1)                   
 rmarkdown     1.11       2018-12-08 [1] CRAN (R 3.5.1)                   
 rprojroot     1.3-2      2018-01-03 [1] CRAN (R 3.5.1)                   
 rstudioapi    0.8        2018-10-02 [1] CRAN (R 3.5.1)                   
 search        0.1.0      2018-12-12 [1] local                            
 sessioninfo   1.1.0      2018-09-25 [1] CRAN (R 3.5.1)                   
 shiny         1.1.0      2018-05-17 [1] CRAN (R 3.5.1)                   
 storage       0.1.0      2018-11-28 [1] local                            
 stringi       1.2.4      2018-07-20 [1] CRAN (R 3.5.1)                   
 stringr       1.3.1      2018-05-10 [1] CRAN (R 3.5.1)                   
 testthat      2.0.0.9000 2018-11-05 [1] local                            
 tibble        1.4.2      2018-01-22 [1] CRAN (R 3.5.1)                   
 tidyr         0.8.2      2018-10-28 [1] CRAN (R 3.5.1)                   
 tidyselect    0.2.5      2018-10-11 [1] CRAN (R 3.5.1)                   
 ui            0.1        2018-12-13 [1] local                            
 usethis       1.4.0      2018-08-14 [1] CRAN (R 3.5.1)                   
 utilities     0.1        2018-12-04 [1] local                            
 withr         2.1.2      2018-03-15 [1] CRAN (R 3.5.1)                   
 xfun          0.4        2018-10-23 [1] CRAN (R 3.5.1)                   
 xtable        1.8-3      2018-08-29 [1] CRAN (R 3.5.1)                   
 yaml          2.2.0      2018-07-25 [1] CRAN (R 3.5.1)  
```

**Comment by yihui** [comment]
You are right that this issue was caused by 0d9f1c4cd3a124088723ff30050e19f62f3d54ef. That particular fix was correct, but the code on the next line is buggy.

One workaround is to store the path in a variable, e.g.,

```r
s <- system.file('scripts/iris-model.R', package = 'repository')
knitr::read_chunk(s)
```

but I guess there will be a new problem. Let me think more about it.

**Comment by yihui** [comment]
Actually could you tell me how to reproduce this issue? I cannot think of a reason why it could occur. I mean `R CMD build/check` shouldn't actually call `process_tangle.block`.

**Comment by lbartnik** [comment]
I can work on a repro, but maybe this will help you figure it our sooner. I did this as a workaround:

```r
rc <- knitr::read_chunk
rc(system.file("....", pacakge = "..."))
```

and it worked. It seems to me that you evaluate the call in an environment that has no access to knitr's "workspace" environment.

Here's the full file where I found the bug, line 137: https://github.com/lbartnik/chronicler/blob/master/vignettes/introduction.Rmd
And here's the fixing commit: https://github.com/lbartnik/chronicler/commit/e5e4550b44aee76dcae88fee9c0654089ff4a37e#diff-b685d909d94516831885198e9c0caba9
